### PR TITLE
Compression

### PR DESCRIPTION
**Pull request by thekid**
This pull request implements the feature suggested in #21 using https://github.com/xp-forge/compression. 

## TL;DR

By default, you don't have to do anything. If your PHP setup supports compression (*Windows PHP comes with zlib enabled by default [1], Ubuntu's `php-cli` package's default configuration also includes it*), this library will signal it can handle it, and - should the server then decide to deliver compressed results - take care of uncompressing the results transparently.

## BC breaks

However, **if** you've been manually handling compression by adding an *Accept-Encoding* header using `with()` and looking at the *Content-Encoding* response header for reading compressed data, your code will stop working: `Response::stream()` will already return a compressed input stream, wrapping that inside another will not work. **The manual handling code will need to be removed when upgrading!**

* * * 

## Sending `Accept-Encoding`

Indicates the content encoding that the client can understand, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding. This can be accomplished as follows:

```php
use webservices\rest\Endpoint;
use io\streams\Compression;

// Detect supported compression algorithms and set "Accept-Encoding" accordingly
// This will typically yield "gzip", see above.
$endpoint= new Endpoint($api);

// Send "Accept-Encoding: identity", indicating the server should not compress
$endpoint= (new Endpoint($api))->compressing(Compression::$NONE);

// Send "Accept-Encoding: gzip, br"
$endpoint= (new Endpoint($api))->compressing(['gzip', 'br']);

// Do not send an "Accept-Encoding" header, i.e. no preference is expressed
// This is the same as the previous behavior
$endpoint= (new Endpoint($api))->compressing(null);
```

## Handling `Content-Encoding`

Indicates the content encoding that the server chose based on the above, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding. This header is handled automatically.

```php
use webservices\rest\Endpoint;

// Even though we don't set an "Accept-Encoding" header, the response may still
// be compressed - omitting it is the same as "*", or "anything is acceptable".
$endpoint= (new Endpoint($api))->compressing(null);
$stream= $endpoint->resource('...')->get();

// io.streams.compress.GzipInputStream instance for "Content-Encoding: gzip"
// io.streams.compress.BrotliInputStream instance for "Content-Encoding: br"
// peer.http.HttpInputStream for uncompressed responses
// 
// In any case, reading works exactly the same
while ($stream->available()) {
  $chunk= $stream->read();
}
```

1: https://www.php.net/manual/en/zlib.installation.php

### CODE REVIEW

**Comment by thekid** [comment]
I fear calling *compressing()* will become boilerplate, and all instantiations will look like this:

```php
// Detect which compression algorithms are supported and set the
// "Accept-Encoding" header accordingly
$endpoint= (new Endpoint($api))->compressing();
```

Maybe detecting supported algorithms should be the default?

**Comment by thekid** [comment]
> Maybe detecting supported algorithms should be the default?

This would be a BC break, we would be sending headers we previously didn't...

**Comment by thekid** [comment]
> This would be a BC break, we would be sending headers we previously didn't...

If we go for a major version, we could do the following:

```php
use webservices\rest\Endpoint;
use io\streams\Compression;

// Detect supported compression algorithms and set "Accept-Encoding" accordingly
// This will typically yield "gzip": Windows PHP comes with zlib enabled by 
// default [1], Ubuntu's php-cli package's default configuration includes it
$endpoint= new Endpoint($api);

// Send "Accept-Encoding: identity", indicating the server should not compress
$endpoint= (new Endpoint($api))->compressing(Compression::$NONE);

// Send "Accept-Encoding: gzip, br"
$endpoint= (new Endpoint($api))->compressing(['gzip', 'br']);

// Do not send an "Accept-Encoding" header, i.e. no preference is expressed
// This is the same as the previous behavior
$endpoint= (new Endpoint($api))->compressing(null);
```
